### PR TITLE
Introducing HookedEncoder

### LINKED ISSUE #258: [Proposal] Support BERT Model Series (and/or T5).
Screenshot 2023-04-26 at 8 00 40 pm

### Proposal 

A glaring gap in models supported by TransformerLens is the lack of support for encoder/decoder models like BERT and T5. In order to rectify this, we need to build support for encoder-decoder models, cross attention and likely a few other details. 

### Motivation

- I've been asked a number of times by people whether we have T5 or BERT support coming. 
- I think this is a reasonable contribution for people skilling up. You can ask for help or collaborate with other people working on building it. 
- I (Joseph) and likely others will be able to give feedback on your work and pitch in if needed! 

### Pitch

In order to complete this card we are going to build a class called HookedEncoderDecoder which can be instantiated and have the BERT or T5 weights from [huggingface](https://huggingface.co/bert-base-uncased) loaded in. This is the same strategy the package uses for other models so there's nothing fundamentally new, only components we haven't yet built with hooks. Once we have a class that has corresponding weights as needed, we will want to ensure all components are "hooked" correctly, enabling methods like .run_with_cache to work. It's not as complicated as it sounds to get it working. 

Possibly the "best" version of this is tricky but we're willing to tolerate some code duplication as long we're confident the implementation is accurate (ie: tests show the hooked activations are correct and inference is identical to what we get without using TransformerLens. 

Tasks:

- [ ] Make a fork for BERT support (keep it synced if building this takes more than a moment)
- [ ] Write some tests to be used during development, these should cover any novel architecture components these will involve during development. I'd start with two placeholder tests. 1. for do you have a model architecture which you can load weights into. and 2. if you had that architecture, would it produce identical results to the hugging-face model. 
- [ ] Load the BERT model, look at each component. Maybe read the BERT paper if needed to get context. Make a list of every nn.module that we don't currently support. Some items that will be on the list. 1. a linear block before the unembed, post-layerNorm (can't be unfolded RIP), sinusoidal positional embeddings. T5 might have other stuff.
- [ ] Keep in mind you might need a new config class to configure the entire architecture. If so, build it. 
- [ ] If you can pass the test for the model doing inference correctly and show that you can use .run_activation_cache and all the other methods that work on hooked transformer (within reason), then we're likely ready to merge. 

Some things to note:
- You should refine the to do list and add detail as you go. 
- Writing tests should make your life easier and create feedback cycles. It will also make it easier if multiple people want to collaborate on this. 
- A [previous PR ](https://github.com/neelnanda-io/TransformerLens/pull/82) attempted to provide BERT support is here. Unfortunately with drift it was too complicated to Merge. To avoid this, I recommend either syncing regularly or getting someone else to finish the PR if you can't. 
- Possibly quite useful, [a link to the original ARENA materials where we did this](https://github.com/callummcdougall/arena-v1/blob/main/w2d2/solutions_build_bert.py) without the hooks during the bootcamp

### Additional context

- [BERT paper](https://arxiv.org/abs/1810.04805)
- [BERT on huggingface](https://huggingface.co/bert-base-uncased)
- [Related T5 Model issue](https://github.com/neelnanda-io/TransformerLens/issues/94)

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/Farama-Foundation/MiniGrid/issues) in the repo (**required**)

### PR DESCRIPTION
**Pull request by rusheb**
# Description

***This feature was co-authored with @MatthewBaggins. Thanks also to @luciaquirke and @jbloomAus for helpful discussions throughout.***

Introducing **`HookedEncoder`**, a BERT-style encoder inheriting from `HookedRootModule`. Weights can be loaded from the huggingface [bert-base-cased](https://huggingface.co/bert-base-cased/tree/main) pretrained model. 

Unlike `HookedTransformer`, it does not (yet) do any pre-processing of the weights (e.g. folding LayerNorm). Another difference is that the model can currently only be run with tokens, and not strings or list of strings. Currently, the supported task/architecture is masked language modelling. Next sentence prediction, causal language modelling, and other tasks are not supported. The HookedEncoder does not contain dropouts, which may lead to inconsistent results when pretraining.

Though this implementation is not fully-featured, I think it's a good starting point which is useful and which we can iterate on. I've tried to keep the scope as small as possible for a number of reasons
  - Getting earlyish feedback on this PR
  - Getting earlier user feedback
  - Reducing the risk of getting distracted by other priorities. (Personally, I won't have much time to work on this over the next two months.)

## Notes
- In the end, based on the discussion in #262, I decided to reuse HookedTransformerConfig rather than creating a new class HookedEncoderConfig. I believe the configs could still be separated if the need arises.
- I chose the masked language modelling task based on feedback from people who want to do research using BERT.

## Key uncertainties
- I'm really not sure about the naming. Currently I have named the main transformer class `HookedEncoder`, but I've prefixed the new component names with `Bert`.
- I'm hopeful that this will be easy to extend to support more variations on the architecture, like HookedTransformer, but I'd like to hear if anybody thinks otherwise.

## Summary of Changes
- Add new class HookedEncoder
- Add new components
	- TokenTypeEmbed
	- BertEmbed
	- BertMLMHead
	- BertBlock
 - Add `additive_attention_mask` parameter to forward method of `Attention` component
 - Add BERT config and state dict to loading_from_pretrained
- Extract methods from HookedTransformer for reuse: 
	- devices.move_to_and_update_config
	- lodaing.fill_missing_keys
- Add demo notebook `demos/BERT.ipynb`
- Update Available Models list in Main Demo
- Testing
  - Unit and acceptance tests for HookedEncoder and sub-components
  - New demo in `demos/BERT.ipynb` also acts as a test
  - I also added some tests for existing components e.g. HookedTransformerConfig

## Future work
- Add support for different tasks, e.g. Next Sentence Prediction, Causal Language Modelling
- Add more models: bert-base-uncased, bert-large-cased, bert-large-uncased
- Add preprocessing of weights including LayerNorm folding
- Accept strings as input and add tokenization helpers from HookedTransformer
- Add support for training/finetuning (most notably, dropouts)
- Adding examples of research using BERT to the demo notebooks

Fixes #258 

# Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### CODE REVIEW

File: transformer_lens/components.py [hunk 1]
```diff
@@ -90,6 +90,98 @@ def forward(
...
+
+    def __init__(self, cfg: Union[Dict, HookedTransformerConfig]):
+        super().__init__()
+        if isinstance(cfg, Dict):
+            cfg = HookedTransformerConfig.from_dict(cfg)
+        self.cfg = cfg
+        self.W_token_type = nn.Parameter(torch.empty(2, self.cfg.d_model))
+
+    def forward(self, token_type_ids: Int[torch.Tensor, "batch pos"]):
+        # TODO: should this method take the tokens rather than the token_type_ids?
```
Comment (rusheb) [inline transformer_lens/components.py:108]: Remove this

**Comment by jbloomAus** [comment]
Thanks @rusheb  and @MatthewBaggins ! This is stellar work! I can't wait to see subsequent investigations!
